lib/collection/src/shards/transfer/resharding_stream_records.rs

```diff
@@ -80,8 +81,51 @@ pub(crate) async fn transfer_resharding_stream_records(
                 "Shard {shard_id} not found"
             )));
         };
-        let transfer_size =
-            (count_result.count as f32 * hashring.resharding_transfer_fraction()) as usize;
+
+        // Resharding up:
+        //
+        // - shards: 1 -> 2
+        //   points: 100 -> 50/50
+        //   transfer points of each shard: 50/1 = 50 -> 50/100 = 50%
+        //   transfer fraction to each shard: 1/new_shard_count = 1/2 = 0.5
+        // - shards: 2 -> 3
+        //   points: 50/50 -> 33/33/33
+        //   transfer points of each shard: 33/2 = 16.5 -> 16.5/50 = 33%
+        //   transfer fraction to each shard: 1/new_shard_count = 1/3 = 0.33
+        // - shards: 3 -> 4
+        //   points: 33/33/33 -> 25/25/25/25
+        //   transfer points of each shard: 25/3 = 8.3 -> 8.3/33 = 25%
+        //   transfer fraction to each shard: 1/new_shard_count = 1/4 = 0.25
+        //
+        // Resharding down:
+        //
+        // - shards: 2 -> 1
+        //   points: 50/50 -> 100
+        //   transfer points of each shard: 50/1 = 50 -> 50/50 = 100%
+        //   transfer fraction to each shard: 1/new_shard_count = 1/1 = 1.0
+        // - shards: 3 -> 2
+        //   points: 33/33/33 -> 50/50
+        //   transfer points of each shard: 33/2 = 16.5 -> 16.5/33 = 50%
+        //   transfer fraction to each shard: 1/new_shard_count = 1/2 = 0.5
+        // - shards: 4 -> 3
+        //   points: 25/25/25/25 -> 33/33/33
+        //   transfer points of each shard: 25/3 = 8.3 -> 8.3/25 = 33%
+        //   transfer fraction to each shard: 1/new_shard_count = 1/3 = 0.33
+        let new_shard_count = match &hashring {
+            HashRingRouter::Single(_) => {
+                return Err(CollectionError::service_error(format!(
+                    "Failed to do resharding transfer, hash ring for shard {shard_id} not in resharding state",
+                )));
+            }
+            HashRingRouter::Resharding { old, new } => {
+                debug_assert!(
+                    old.len().abs_diff(new.len()) <= 1,
+                    "expects resharding to only move up or down by one shard",
+                );
+                new.len()
+            }
+        };
+        let transfer_size = (count_result.count as f64 / new_shard_count as f64) as usize;
```

Comment:
Just do integer division without `f64` conversion? We are converting back into `usize` anyway?